File: dynabuffers-python/dynabuffers/api/map/DynabuffersMap.py
from typing import List

from dynabuffers.api.ISerializable import ISerializable
from dynabuffers.ast.ClassType import ClassType
from dynabuffers.ast.UnionType import UnionType
from dynabuffers.ast.datatype.OptionType import OptionType
from dynabuffers.ast.datatype.RefType import RefType


class DynabuffersMap(dict):

    def __init__(self, map: dict, tree: List[ISerializable], root: ISerializable):
        super().__init__()
        self.raw_map = map
        self.update(map)
        self.tree = tree
        self.clazz = self.resolve_class(tree, root, map)

    def __getitem__(self, key):
        if not self.is_defined(key):
            raise ValueError("field " + str(key) + " is not defined in " + str(self.clazz.options.name))

        result = self.raw_map[key]
        if result is None and not self.is_optional(key):
            raise ValueError("field " + str(key) + " is not optional in " + str(self.clazz.options.name))

        if result is not None and self.is_ref_type(key):
            ref_type = self.get_ref_type(key)
            if ref_type is not None:
                return DynabuffersMap(result, self.tree, ref_type)

        return result

    def get_ref(self, key: str) -> 'DynabuffersMap':
        if isinstance(self[key], dict):
            return DynabuffersMap(self[key], self.tree, self.clazz)
        raise ValueError("field " + str(key) + " with result " + str(self[key]) + " is not a ref type")

    def get_string(self, key: str) -> str:
        if isinstance(self[key], str):
            return self[key]
        raise ValueError("field " + str(key) + " with result " + str(self[key]) + " is not a string type")

    def get_int(self, key: str) -> int:
        if isinstance(self[key], int):
            return self[key]
        raise ValueError("field " + str(key) + " with result " + str(self[key]) + " is not a int type")

    def get_short(self, key: str) -> int:
        if isinstance(self[key], int):
            return self[key]
        raise ValueError("field " + str(key) + " with result " + str(self[key]) + " is not a short type")

    def get_byte(self, key: str) -> int:
        if isinstance(self[key], int):
            return self[key]
        raise ValueError("field " + str(key) + " with result " + str(self[key]) + " is not a byte type")

    def get_long(self, key: str) -> int:
        if isinstance(self[key], int):
            return self[key]
        raise ValueError("field " + str(key) + " with result " + str(self[key]) + " is not a long type")

    def get_float(self, key: str) -> float:
        if isinstance(self[key], float):
            return self[key]
        raise ValueError("field " + str(key) + " with result " + str(self[key]) + " is not a float type")

    def get_map(self, key: str) -> int:
        if isinstance(self[key], dict):
            return self[key]
        raise ValueError("field " + str(key) + " with result " + str(self[key]) + " is not a map type")

    def get_bytearray(self, key: str) -> bytes:
        if isinstance(self[key], bytes):
            return self[key]
        raise ValueError("field " + str(key) + " with result " + str(self[key]) + " is not a bytearray type")

    def resolve_class(self, tree: List[ISerializable], root: ISerializable, source: dict):
        if isinstance(root, ClassType):
            return root
        if isinstance(root, UnionType):
            if source[":type"] is None:
                return list(filter(lambda x: x.options.name == root.options.values[0], tree))[0]

            type = source[":type"] if isinstance(source[":type"], str) else root.options.values[source[":type"]]
            return list(filter(lambda x: x.options.name == type, tree))[0]
        raise ValueError("cannot resolve class from " + str(root))

    def is_defined(self, key: str) -> bool:
        if key == ":type":
            return True
        return next((x for x in self.clazz.options.fields if x.options.name == key), False)

    def is_optional(self, key: str) -> bool:
        if key == ":type":
            return False
        field = next((x for x in self.clazz.options.fields if x.options.name == key), None)
        return isinstance(field.options.dataType, OptionType) if field is not None else False

    def is_ref_type(self, key: str) -> bool:
        if key == ":type":
            return False
        field = next((x for x in self.clazz.options.fields if x.options.name == key), None)
        return isinstance(field, RefType)

    def get_ref_type(self, key: str) -> ClassType:
        field = next((x for x in self.clazz.options.fields if x.options.name == key), None)
        name = field.options.dataType.options.name
        type = next((x for x in self.tree if x.options.name == name), False)

        source = self.raw_map[key]
        return self.resolve_class(self.tree, type, source) if isinstance(type, UnionType) else type

File: dynabuffers-java/src/main/kotlin/dynabuffers/api/map/DynabuffersMap.kt
package dynabuffers.api.map

import dynabuffers.SpecialKey
import dynabuffers.api.IType
import dynabuffers.ast.ClassType
import dynabuffers.ast.UnionType
import dynabuffers.ast.datatype.OptionType
import dynabuffers.ast.datatype.RefType
import dynabuffers.exception.DynabuffersException

open class DynabuffersMap(map: Map<String, Any?>,
                          private val tree: List<IType>,
                          root: IType) : HashMap<String, Any?>(map) {

    private val clazz = resolveClass(tree, root)

    override fun getOrDefault(key: String, defaultValue: Any?): Any? {
        if (!isDefined(key))
            throw DynabuffersException("field $key is not defined in ${clazz.options.name}")
        return super.getOrDefault(key, defaultValue)
    }

    override fun get(key: String): Any? {
        if (!isDefined(key))
            throw DynabuffersException("field $key is not defined in ${clazz.options.name}")

        val result = super.get(key)
        if (result == null && !isOptional(key))
            throw DynabuffersException("field $key is not optional in ${clazz.options.name}")

        @Suppress("UNCHECKED_CAST")
        if (result != null && isRefType(key)) {
            val refType = getRefType(key)
            if (refType != null)
                return DynabuffersMap(result as Map<String, Any?>, tree, refType)
        }

        return result
    }

    fun getRef(key: String): DynabuffersMap {
        val result = get(key)
        if (result is DynabuffersMap) return result
        throw DynabuffersException("field $key with result $result is not a ref type")
    }

    fun getString(key: String): String {
        val result = get(key)
        if (result is String) return result
        throw DynabuffersException("field $key with result $result is not of type string")
    }

    fun getInt(key: String): Int {
        val result = get(key)
        if (result is Number) return result.toInt()
        throw DynabuffersException("field $key with result $result is not of type int")
    }

    fun getShort(key: String): Short {
        val result = get(key)
        if (result is Number) return result.toShort()
        throw DynabuffersException("field $key with result $result is not of type short")
    }

    fun getByte(key: String): Byte {
        val result = get(key)
        if (result is Number) return result.toByte()
        throw DynabuffersException("field $key with result $result is not of type byte")
    }

    fun getLong(key: String): Long {
        val result = get(key)
        if (result is Number) return result.toLong()
        throw DynabuffersException("field $key with result $result is not of type long")
    }

    fun getFloat(key: String): Float {
        val result = get(key)
        if (result is Number) return result.toFloat()
        throw DynabuffersException("field $key with result $result is not of type float")
    }

    @Suppress("UNCHECKED_CAST")
    fun getMap(key: String): Map<String, Any?> {
        val result = get(key)
        if (result is Map<*, *>) return result as Map<String, Any?>
        throw DynabuffersException("field $key with result $result is not of type map")
    }

    fun getByteArray(key: String): ByteArray {
        val result = get(key)
        if (result is ByteArray) return result
        throw DynabuffersException("field $key with result $result is not of type [byte]")
    }

    /**
     * Indicates if the given key value is defined.
     */
    private fun isDefined(key: String): Boolean {
        return when (key) {
            ":type" -> true
            SpecialKey.NAMESPACE.key -> true
            else -> clazz.options.fields.any { it.options.name == key }
        }
    }

    /**
     * Indicates if the given key value is optional.
     */
    private fun isOptional(key: String): Boolean {
        if (key == ":type") return false
        val field = clazz.options.fields.find { it.options.name == key }
        return if (field != null) field.options.dataType is OptionType else false
    }

    /**
     * Indicates if the given key value is a RefType.
     */
    private fun isRefType(key: String): Boolean {
        if (key == ":type" || key == SpecialKey.NAMESPACE.key) return false
        val field = clazz.options.fields.find { it.options.name == key }!!
        return field.options.dataType is RefType
    }

    @Suppress("UNCHECKED_CAST")
    private fun getRefType(key: String): ClassType? {
        val field = clazz.options.fields.find { it.options.name == key }!!
        val name = (field.options.dataType as RefType).options.name
        val type = tree.find {
            when (it) {
                is ClassType -> it.options.name == name
                is UnionType -> it.options.name == name
                else -> false
            }
        } ?: return null

        val source = super.get(key) as Map<String, Any?>
        return (if (type is UnionType) resolveClass(tree, type, source) else type) as ClassType
    }

    private fun resolveClass(tree: List<IType>, root: IType, source: Map<String, Any?> = this): ClassType {
        if (root is ClassType) return root
        if (root is UnionType) {
            if (source[":type"] == null) {
                return tree.filterIsInstance<ClassType>()
                        .find { it.options.name == root.options.values[0] }!!
            }
            val type = when (source[":type"]) {
                is String -> source[":type"]
                else -> root.options.values[source[":type"] as Int]
            }
            return tree.filterIsInstance<ClassType>()
                    .find { it.options.name == type }!!
        }
        throw DynabuffersException("cannot resolve class from $root")
    }

}
